### PR TITLE
issue-63565: community requested small QoL fix for more configurabili…

### PR DESCRIPTION
**Pull request by ghost**
This is a small Quality of Life patch. It adds more configurability to `--function-arg-placeholders`.

The current issue for this is: [63565](https://github.com/llvm/llvm-project/issues/63565).  I say current, because more issues where made on this topic. My original issue was: [390](https://github.com/clangd/vscode-clangd/issues/390). Some of my motivations can be read in that thread: [here](https://github.com/clangd/vscode-clangd/issues/390#issuecomment-1647401973) and [here](https://github.com/clangd/vscode-clangd/issues/390#issuecomment-1645383861)

You might want to checkout the outputted help text, which is set in `ClangdMain.cpp`. It is not bad, but perhaps give it a last critical look!?

I'll be awaiting your reply or merger.

Thanks!

### CODE REVIEW

**Review by HighCommander4 — CHANGES_REQUESTED** [review]
Thanks for putting together this patch!

Two high-level pieces of feedback:
 1. The current patch implements the options as additional values of the `--function-arg-placeholders` command-line flag. However, the proposal in [this comment](https://github.com/llvm/llvm-project/issues/63565#issuecomment-1975065771) was to make these values of a new config file key called `ArgumentLists`, under [`Completion`](https://clangd.llvm.org/config.html#completion). (This involves a bit of boilerplate; it helps to look at a previous patch which adds a new config key, such as [this one](https://github.com/llvm/llvm-project/commit/16fd5c278488b3d3275afc381a00ba0b51b070ee).)
 2. It would be great to have some unit tests exercising the options. We have an existing unit test exercising the current boolean flag [here](https://github.com/llvm/llvm-project/blob/587d4cc797e09eeb2ec93d8b4ee9cd484626bf8f/clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp#L2598) which can be extended as appropriate.

**Comment by MK-Alias** [comment]
Thanks for the feedback. Will update the patch soon! (I've got lots of stuff to do, but it's high on my todo-list!)

Just to be clear, do you want to keep the new `--function-arg-placeholders` flags and also add it to the `.clangd` config!? Or do you only want the `.clangd` config and drop the changes to the `--function-arg-placeholders` flag!?

**Comment by HighCommander4** [comment]
> Just to be clear, do you want to keep the new `--function-arg-placeholders` flags and also add it to the `.clangd` config!? Or do you only want the `.clangd` config and drop the changes to the `--function-arg-placeholders` flag!?

The latter. The only change related to the command-line flag that's needed, is to translate its (still boolean) value into the new (enum-valued) config option; we have a class named [FlagsConfigProvider](https://searchfox.org/llvm/rev/36ad0720de623221e3cc17d30f4173331c099a72/clang-tools-extra/clangd/tool/ClangdMain.cpp#638) which helps with that (so that the rest of the code only needs to care about the config option).

**Comment by MK-Alias** [comment]
Thanks for the input, a followup question is: Which one takes precedence? Does the `--function-arg-placeholders` argument take precedence over `.clangd` config!? Or the other way around!?

**Comment by HighCommander4** [comment]
> Thanks for the input, a followup question is: Which one takes precedence? Does the `--function-arg-placeholders` argument take precedence over `.clangd` config!? Or the other way around!?

`.clangd` should take precedence because it's the a more specific setting (command line arguments apply to the whole clangd process, `.clangd` files are scoped to a directory).

If you are handling the flag in `FlagsConfigProvider` as suggested, this will be the resulting behaviour will no further logic needed.

**Comment by MK-Alias** [comment]
Because the `.clangd` config is made out of blocks, and no block was specified for our new `ArgumentLists` key, I've put it under `Completion`. So now it is like
```yaml
Completion:
  AllScopes: None
  ArgumentLists: None
```

It seems like the most logical choice, because it's an auto-completion feature!

Is this okay!? Or must it go in an other block!?

**Comment by HighCommander4** [comment]
> Is this okay!?

Yep. By the way, this was specified in the [original proposal](https://github.com/llvm/llvm-project/issues/63565#issuecomment-1975065771):

> Here's a concrete proposal for a config file syntax that would address this request, https://github.com/clangd/clangd/issues/1252, and supersede `--function-arg-placeholders`: add a new key **under `Completion`** called `ArgumentLists`, with the following supported values:

**Comment by MK-Alias** [comment]
Updated the source. It should now confirm to everything. To recap: `.clangd` now has a added key: `ArgumentLists` in the `Completion` block. It supports these values: `None` | `OpenDelimiter` | `Delimiters` | `FullPlaceholders`. The `.clangd` config will overwite the prevously used: `--function-arg-placeholders`, which means that the only effect still done by `--function-arg-placeholders=0`, is when `.clangd` has no `ArgumentLists` value. In which case it will use the `Delimiters`.

UnitTest have also been updated. Other then the tests already in place, it will also check the `None` option and the `OpenDelimiter` option. The `FullPlaceholders` option is not checked. It was also never checked prior to this patch.

As mentions in [this post](https://github.com/llvm/llvm-project/issues/63565#issuecomment-2337396222): these to issues remain within vsocde/vscode-clangd:

- None: after completion an error squiggly line will occur on the identifier. Would be slightly nicer if it wouldn't.
- OpenDelimiter: the signature help hover doesn't spawn. It seems that [Sam's Patch from #398](https://github.com/clangd/vscode-clangd/pull/398/files) needs to be reconfigured for this.

The OpenDelimiter issue seems like a small fix and would generally improve the UX on this. 
    

Ill be awaiting your response.

File: clang-tools-extra/clangd/ConfigCompile.cpp [hunk 1]
```diff
@@ -504,10 +504,10 @@ struct FragmentCompiler {
       auto Fast = isFastTidyCheck(Str);
       if (!Fast.has_value()) {
         diag(Warning,
-             llvm::formatv(
-                 "Latency of clang-tidy check '{0}' is not known. "
-                 "It will only run if ClangTidy.FastCheckFilter is Loose or None",
-                 Str)
+             llvm::formatv("Latency of clang-tidy check '{0}' is not known. "
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/ConfigCompile.cpp:507]: nit: I'm seeing some diff hunks containing unrelated formatting changes. If would be great to remove this as they pollute the blame

File: clang-tools-extra/clangd/ConfigCompile.cpp [hunk 2]
```diff
@@ -622,6 +622,31 @@ struct FragmentCompiler {
             C.Completion.AllScopes = AllScopes;
           });
     }
+    if (F.ArgumentLists) {
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/ConfigCompile.cpp:625]: We have a `compileEnum` helper to do this more easily, see [the handling of `BackgroundPolicy`](https://searchfox.org/llvm/rev/87d56c59f52d033cd7c46d769338b9c47fea4929/clang-tools-extra/clangd/ConfigCompile.cpp#326-334) as an example

File: clang-tools-extra/clangd/ConfigFragment.h [hunk 3]
```diff
@@ -308,6 +309,13 @@ struct Fragment {
     /// Whether code completion should include suggestions from scopes that are
     /// not visible. The required scope prefix will be inserted.
     std::optional<Located<bool>> AllScopes;
+    /// How to present the argument list between '()' and '<>':
+    /// valid values are enum Config::ArgumentListsOption values:
+    ///   None: Nothing at all
+    ///   OpenDelimiter: only opening delimiter "(" or "<"
+    ///   Delimiters: empty pair of delimiters "()" or "<>"
+    ///   FullPlaceholders: full name of both type and variable
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/ConfigFragment.h:317]: nit: variable --> parameter

File: clang-tools-extra/clangd/ConfigYAML.cpp [hunk 4]
```diff
@@ -230,6 +230,10 @@ class Parser {
       if (auto AllScopes = boolValue(N, "AllScopes"))
         F.AllScopes = *AllScopes;
     });
+    Dict.handle("ArgumentLists", [&](Node &N) {
+      if (auto ArgumentLists = scalarValue(N, "ArgumentLists"))
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/ConfigYAML.cpp:234]: since `F.ArgumentLists` is itself an `optional`, this can be simplified to:

`F.ArgumentLists = scalarValue(N, "ArgumentLists");`

File: clang-tools-extra/clangd/tool/ClangdMain.cpp [hunk 5]
```diff
@@ -242,13 +242,16 @@ opt<std::string> FallbackStyle{
...
 };
 
-opt<bool> EnableFunctionArgSnippets{
-    "function-arg-placeholders",
-    cat(Features),
-    desc("When disabled, completions contain only parentheses for "
-         "function calls. When enabled, completions also contain "
-         "placeholders for method parameters"),
-    init(CodeCompleteOptions().EnableFunctionArgSnippets),
+opt<Config::ArgumentListsOption> PlaceholderOption{
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/tool/ClangdMain.cpp:245]: For simplicity and consistency with `--background-index`, let's keep the name and type of this unchanged

File: clang-tools-extra/clangd/tool/ClangdMain.cpp [hunk 6]
```diff
@@ -694,13 +698,20 @@ class FlagsConfigProvider : public config::Provider {
...
+    }
+
     Frag = [=](const config::Params &, Config &C) {
       if (CDBSearch)
         C.CompileFlags.CDBSearch = *CDBSearch;
       if (IndexSpec)
         C.Index.External = *IndexSpec;
       if (BGPolicy)
         C.Index.Background = *BGPolicy;
+      if (ArgumentLists && C.Completion.ArgumentLists ==
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/tool/ClangdMain.cpp:712]: For reasons mentioned above, no need for the second condition here (at this point, the config provider for the config file has not **yet** assigned anything to `C`; when it does run later, it will overwrite the field if needed)

File: clang-tools-extra/clangd/CodeComplete.cpp [hunk 7]
```diff
@@ -561,14 +561,32 @@ struct CodeCompletionBuilder {
   }
 
   std::string summarizeSnippet() const {
+
+    /// localize a ArgList depending on the config. If the config is unset we
+    /// will use our local (this) version, else use the one of the config
+    const Config::ArgumentListsOption ArgList =
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/CodeComplete.cpp:567]: For consistency with our existing `Completion` option (`AllScopes`), let's do the `Config` lookup [here](https://searchfox.org/llvm/rev/87d56c59f52d033cd7c46d769338b9c47fea4929/clang-tools-extra/clangd/ClangdServer.cpp#453) and propagate it into `CodeCompleteOpts`.

That then gets propagated to `CodeCompletionBuilder::ArgumentLists` in the `CodeCompletionBuilder` constructor, and here we can just use the `ArgumentLists` member without doing anything further.

File: clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp [hunk 8]
```diff
@@ -6,6 +6,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "../Config.h"
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp:9]: This can be included without the `../`. (Please alphabetize it among the existing includes.)

File: clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp [hunk 9]
```diff
@@ -2595,10 +2596,10 @@ TEST(SignatureHelpTest, DynamicIndexDocumentation) {
               ElementsAre(AllOf(sig("foo() -> int"), sigDoc("Member doc"))));
 }
 
-TEST(CompletionTest, CompletionFunctionArgsDisabled) {
+TEST(CompletionTest, ArgumentListsNotFullPlaceholders) {
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp:2599]: nit: suggestion for slightly more general test name: `ArgumentListsPolicy`

**Review by HighCommander4 — CHANGES_REQUESTED** [review]
Thanks! Generally looks pretty good, a few minor comments / cleanups and this should be good to go.

File: clang-tools-extra/clangd/Config.h [hunk 10]
```diff
@@ -126,11 +126,29 @@ struct Config {
...
+  enum class ArgumentListsOption {
+    /// the default value. This will imply FullPlaceholders unless overridden by
+    /// --function-arg-placeholders=0, in which case Delimiters is used.
+    /// any other use-case of --function-arg-placeholders is ignored
+    UnsetDefault = 0,
+    /// nothing, no argument list and also NO Delimiters "()" or "<>"
+    None,
+    /// open, only opening delimiter "(" or "<"
+    OpenDelimiter,
+    /// empty pair of delimiters "()" or "<>" (or [legacy] alias 0)
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/Config.h:139]: nit: these mentions of legacy aliases can be removed (it's just confusing since that applies to the command line flag which is defined elsewhere)

**Comment by HighCommander4** [comment]
> As mentions in [this post](https://github.com/llvm/llvm-project/issues/63565#issuecomment-2337396222): these to issues remain within vsocde/vscode-clangd:
> 
>    * None: after completion an error squiggly line will occur on the identifier. Would be slightly nicer if it wouldn't.
> 
>    * OpenDelimiter: the signature help hover doesn't spawn. It seems that [Sam's Patch from #398](https://github.com/clangd/vscode-clangd/pull/398/files) needs to be reconfigured for this.

Could you file a new issue for each of these please?

File: clang-tools-extra/clangd/ConfigCompile.cpp (hunk 1 above)
Comment (MK-Alias) [reply clang-tools-extra/clangd/ConfigCompile.cpp:507]: I use `clang-format` with the `.clang-format` of the code base. Even if I don't change anything and just do a `clang-format` this change will happen. It seems that it's not correctly formatted. I will fix this and my recommendation would be to use `clang-format` on this file - or perhaps the whole code base - post patching in its own commit. this will fix this issue(s) in the future.

File: clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp (hunk 8 above)
Comment (MK-Alias) [reply clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp:9]: >This can be included without the ../

Yes, but it of-course shouldn't!  Confirm the [standard](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1570.pdf#page=182) it should be either: `#include "../Config.h"` or `#include <Config.h>` (and also it should be `Config.hpp`)

But because this is used in the whole codebase, if changed it to `#include "Config.h"`

**Comment by MK-Alias** [comment]
Here a quick summery of the most important things.

The reason I've implemented a `UnsetDefault` is because the program flow doesn't work as [you specify](https://github.com/llvm/llvm-project/pull/108005#discussion_r1764069398). I needed a **untouched** state. Granted that `std::optional` would have been a better choice for this codebase. Thats why I had the [extra check](https://github.com/llvm/llvm-project/pull/108005#discussion_r1764080701) in the `FlagConfigProvider`.

`FlagConfigProvider` is pushed in the vector after `.clangd` config in [CLangdMain.cpp](https://github.com/llvm/llvm-project/blob/a4586bd2d4fa7d6c0100893496a9383fd581e2e9/clang-tools-extra/clangd/tool/ClangdMain.cpp#L926)
Then [ConfigProvider.cpp](https://github.com/llvm/llvm-project/blob/5ac97d397c2088c3ac0a113506e57ab9b1e69ac8/clang-tools-extra/clangd/ConfigProvider.cpp#L154) just iterates over it, which means that `FlagConfigProvider` is evaluated **after** `.clangd` and will override it!

I now have a version which I belief is fully compliant to all your critique., however this means that the `--function-arg-placeholders` will override the `.clangd` `ArgumentLists` setting.

This would be confirm the overlaying logic of the program. So I would recommend just accept this.! If someone ever wants to change the parsing order, our new `ArgumentLists` will just confirm to that.!

I will push the new version in a couple of minutes for you to evaluate.

**Comment by MK-Alias** [comment]
> > As mentions in [this post](https://github.com/llvm/llvm-project/issues/63565#issuecomment-2337396222): these to issues remain within vsocde/vscode-clangd:
> > 
> > * None: after completion an error squiggly line will occur on the identifier. Would be slightly nicer if it wouldn't.
> > * OpenDelimiter: the signature help hover doesn't spawn. It seems that [Sam's Patch from #398](https://github.com/clangd/vscode-clangd/pull/398/files) needs to be reconfigured for this.
> 
> Could you file a new issue for each of these please?

Sure! Will do so when the our changes have been merged!

**Comment by MK-Alias** [comment]
Note: because the command line argument is reverted back to its original boolean: `EnableFunctionArgSnippets` if `--function-arg-placeholders=0` is given, then it will override the `.clangd`, if `--function-arg-placeholders=1` not. This is because the `false` state of `EnableFunctionArgSnippets` is tested in `FlagConfigProvider`

File: clang-tools-extra/clangd/ConfigYAML.cpp (hunk 4 above)
Comment (HighCommander4) [reply clang-tools-extra/clangd/ConfigYAML.cpp:234]: (this comment remains to be addressed)

**Comment by HighCommander4** [comment]
> `FlagConfigProvider` is pushed in the vector after `.clangd` config in [CLangdMain.cpp](https://github.com/llvm/llvm-project/blob/a4586bd2d4fa7d6c0100893496a9383fd581e2e9/clang-tools-extra/clangd/tool/ClangdMain.cpp#L926) Then [ConfigProvider.cpp](https://github.com/llvm/llvm-project/blob/5ac97d397c2088c3ac0a113506e57ab9b1e69ac8/clang-tools-extra/clangd/ConfigProvider.cpp#L154) just iterates over it, which means that `FlagConfigProvider` is evaluated **after** `.clangd` and will override it!

Thanks for catching that; I was under the mistaken impression that it was the other way around!

I think it would be a nicer user experience if the config file took precedence over compile flags... but since that's not the case today for existing flags (like `--background-index`), I think it's appropriate for this patch to remain consistent with the existing behaviour.

In other words, I think what the current version of the patch is doing is fine.

**Review by HighCommander4 — CHANGES_REQUESTED** [review]
Thanks, this is looking pretty good. My only remaining request is to please split the formatting (whitespace) changes out. I know it's the fault of the code not being clang-format clean, and I'm happy to merge them in a separate PR, but it confuses things when "git blame" says that the last commit that touched some unrelated test / code line is this one about `ArgumentLists`.

**Comment by MK-Alias** [comment]
> I think it would be a nicer user experience if the config file took precedence over compile flags...

The whole idea of command-line arguments is to override defaults at execution time. Config-files should be persistent part of the config, and to override the current config, you can use - non persistent - command-line arguments... That's the whole idea behind it is so you can run a command with a _certain intent_, The other way around will render the command line useless, or might have unexpected behavior. (say: supplying a command line argument and it then not doing anything)

> (this comment remains to be addressed)

I really don't know what you mean [here](https://github.com/llvm/llvm-project/pull/108005#discussion_r1769703751)!? can you elaborate? what do you want me to do?

I'm pretty happy with the patch, as-is. Only the boolean on the command-line `EnableFunctionArgSnippets` is [lacking features](https://github.com/llvm/llvm-project/pull/108005#issuecomment-2358650558). I'm checking the `false` state of it, which means that the `true` state is ignored. I've now changed it to a `int` value with the default value of `-1`. So if the value is `-1`, i can ignore it in the `FlagsConfigProvider`. So now both `false` (0) and `true` (1) are properly handled.

> but it confuses things when "git blame" says that the last commit that touched some unrelated....

And it's not confusing if "git blame" says that this same part of code is there in a "clang-format commit". Instead of it's original? The world is simply not perfect. The git blame is going to change these lines anyways.

[A choice have been made to use clang-format](https://llvm.org/docs/Contributing.html#how-to-submit-a-patch). The source currently is not properly formatted because the previous commits where not properly formatted, or a newer version of clang-format has a different interpretations of how to format the code. Me now having to first format the code to comply to the clang-format requirement, and then having to unformat other parts of the source is a absolute annoyance! And also the reason why the [clang-format check](https://github.com/llvm/llvm-project/actions/runs/10924307640/job/30480533009?pr=108005) is now failing. The irony is that people choose to use a code-formatter to not have to discuss or deal with code formatting, and then we now have to discuss the code formatting.

Code-formatters have their pros & cons, and this one of the cons, and personally I would argue that this issue is just "collateral damage" of using a code-formatter. Things like these can and will happen from time to time, if the code isn't properly formatted, or the formatter [clang-format] gets updated. It's simply inherit of the choice of using a formatter.

I will commit the newest version of the patch, which will have the mentioned int value on `EnableFunctionArgSnippets`. It will be properly clang-formatted. And if you have no other critic/insights and agree with merging this version you wave three options.

A: Merge it as is (recommended: this follows the [guide](https://llvm.org/docs/Contributing.html#how-to-submit-a-patch))
B: Clang-format the code-base (or at least the affected source files in this patch) and merge that. And then merge this patch. If there are any conflicts then - there will probably not be, because my version is clang-formatted - I will happily pull the new version and make a new patch if necessary!
C: make a local copy of my patch, reconfigure it to your licking then merge it.

Effected files in this PR are:
```
clang-tools-extra/clangd/ClangdServer.cpp
clang-tools-extra/clangd/CodeComplete.cpp
clang-tools-extra/clangd/CodeComplete.h
clang-tools-extra/clangd/Config.h
clang-tools-extra/clangd/ConfigCompile.cpp
clang-tools-extra/clangd/ConfigFragment.h
clang-tools-extra/clangd/ConfigYAML.cpp
clang-tools-extra/clangd/tool/ClangdMain.cpp
clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp
```

So if you which to first clang-format before merging, you technically only have to do these.

**Comment by HighCommander4** [comment]
> > but it confuses things when "git blame" says that the last commit that touched some unrelated....
> 
> And it's not confusing if "git blame" says that this same part of code is there in a "clang-format commit". Instead of it's original?

I would say that's less confusing, because you know you can just skip that sort of commit in the blame.

I'm not trying to be difficult here; "please remove unrelated formatting changes" is something I've been told by other reviewers in response to my own patches.

**Comment by Darkproduct** [comment]
This is too bad. I was just looking for this feature.

**Comment by HighCommander4** [comment]
I've resubmitted this as https://github.com/llvm/llvm-project/pull/111322.
